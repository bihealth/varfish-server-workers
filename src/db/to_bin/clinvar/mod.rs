//! Code for converting ClinVar database to binary.

use std::{fs::File, io::Write, path::Path, time::Instant};

use prost::Message;
use thousands::Separable;

use crate::{
    common::{build_chrom_map, open_read_maybe_gz, trace_rss_now},
    sv::query::clinvar::pbs::{SvDatabase, SvRecord},
};

mod input;

/// Helper to convert VCV IDs to numbers.
fn numeric_vcv_id(raw_id: &str) -> Result<u32, anyhow::Error> {
    let clean_id: String = raw_id
        .chars()
        .skip("VCV".len())
        .skip_while(|c| *c == '0')
        .collect();
    clean_id
        .parse::<u32>()
        .map_err(|e| anyhow::anyhow!("could not parse VCV id {:?}: {}", &clean_id, &e))
}

/// Perform conversion to protocolbuffers `.bin` file.
pub fn convert_to_bin<P, Q>(path_input_tsv: P, path_output_bin: Q) -> Result<(), anyhow::Error>
where
    P: AsRef<Path>,
    Q: AsRef<Path>,
{
    let chrom_map = build_chrom_map();

    let mut reader = csv::ReaderBuilder::new()
        .comment(Some(b'#'))
        .delimiter(b'\t')
        .has_headers(false)
        .from_reader(open_read_maybe_gz(path_input_tsv)?);
    let before_parsing = Instant::now();

    let mut records = Vec::new();
    for record in reader.deserialize() {
        match record {
            Err(e) => {
                tracing::warn!("error parsing ClinVar record: {}", e);
            }
            Ok(record) => {
                let record: input::Record = record;
                let variation_type: crate::sv::query::clinvar::pbs::VariationType =
                    record.variation_type.into();
                let pathogenicity: crate::sv::query::clinvar::pbs::Pathogenicity =
                    record.pathogenicity.into();
                if let Some(chrom_no) = chrom_map.get(&record.chromosome) {
                    records.push(SvRecord {
                        chrom_no: *chrom_no as i32,
                        start: record.begin + 1,
                        stop: record.end,
                        variation_type: variation_type as i32,
                        pathogenicity: pathogenicity as i32,
                        vcv: numeric_vcv_id(&record.vcv)?,
                    });
                } else {
                    tracing::warn!("unknown chromosome {}", &record.chromosome);
                }
            }
        }
    }
    let clinvar_db = SvDatabase { records };

    tracing::debug!(
        "total time spent reading {} records: {:?}",
        clinvar_db.records.len().separate_with_commas(),
        before_parsing.elapsed()
    );
    trace_rss_now();

    let before_writing = Instant::now();
    let mut output_file = File::create(&path_output_bin)?;
    output_file.write_all(&clinvar_db.encode_to_vec())?;
    output_file.flush()?;
    tracing::debug!(
        "total time spent writing {} records: {:?}",
        clinvar_db.records.len().separate_with_commas(),
        before_writing.elapsed()
    );

    Ok(())
}
