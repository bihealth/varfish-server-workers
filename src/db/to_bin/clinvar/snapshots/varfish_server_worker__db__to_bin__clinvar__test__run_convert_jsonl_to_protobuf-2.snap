---
source: src/db/to_bin/clinvar/mod.rs
expression: records
---
- chrom_no: 6
  start: 101528
  stop: 227833
  variation_type: 2
  pathogenicity: 0
  rcv: 50292
- chrom_no: 0
  start: 147228867
  stop: 147248577
  variation_type: 2
  pathogenicity: 2
  rcv: 50299
- chrom_no: 6
  start: 70356011
  stop: 70537428
  variation_type: 2
  pathogenicity: 2
  rcv: 50317
- chrom_no: 6
  start: 10649696
  stop: 11329787
  variation_type: 2
  pathogenicity: 2
  rcv: 50319
- chrom_no: 20
  start: 35734654
  stop: 35879759
  variation_type: 2
  pathogenicity: 2
  rcv: 50337
- chrom_no: 0
  start: 149041013
  stop: 149699420
  variation_type: 2
  pathogenicity: 0
  rcv: 50339
- chrom_no: 1
  start: 242930600
  stop: 243007359
  variation_type: 2
  pathogenicity: 0
  rcv: 50458
- chrom_no: 5
  start: 259528
  stop: 307998
  variation_type: 2
  pathogenicity: 0
  rcv: 50474
- chrom_no: 5
  start: 259528
  stop: 293493
  variation_type: 2
  pathogenicity: 0
  rcv: 50518
- chrom_no: 2
  start: 2342852
  stop: 2796629
  variation_type: 2
  pathogenicity: 2
  rcv: 50588
- chrom_no: 10
  start: 196966
  stop: 244236
  variation_type: 3
  pathogenicity: 2
  rcv: 50265
- chrom_no: 0
  start: 148867551
  stop: 149699420
  variation_type: 3
  pathogenicity: 0
  rcv: 50266
- chrom_no: 1
  start: 242930600
  stop: 243059659
  variation_type: 3
  pathogenicity: 0
  rcv: 50281
- chrom_no: 22
  start: 155231256
  stop: 155251871
  variation_type: 3
  pathogenicity: 0
  rcv: 50305
- chrom_no: 22
  start: 155231256
  stop: 155251871
  variation_type: 3
  pathogenicity: 0
  rcv: 50306
- chrom_no: 0
  start: 149041013
  stop: 149699420
  variation_type: 3
  pathogenicity: 0
  rcv: 50340
- chrom_no: 4
  start: 629455
  stop: 820475
  variation_type: 3
  pathogenicity: 0
  rcv: 50363
- chrom_no: 11
  start: 61091958
  stop: 61921447
  variation_type: 3
  pathogenicity: 2
  rcv: 50369
- chrom_no: 1
  start: 242930600
  stop: 243007359
  variation_type: 3
  pathogenicity: 0
  rcv: 50463
- chrom_no: 5
  start: 259528
  stop: 307998
  variation_type: 3
  pathogenicity: 0
  rcv: 50482
- chrom_no: 15
  start: 89817659
  stop: 89861760
  variation_type: 3
  pathogenicity: 4
  rcv: 3614
- chrom_no: 7
  start: 53536328
  stop: 53598019
  variation_type: 3
  pathogenicity: 4
  rcv: 4182
- chrom_no: 7
  start: 53537321
  stop: 53574220
  variation_type: 3
  pathogenicity: 4
  rcv: 4183
- chrom_no: 16
  start: 3543558
  stop: 3543559
  variation_type: 1
  pathogenicity: 4
  rcv: 4692
- chrom_no: 4
  start: 77436046
  stop: 77444215
  variation_type: 3
  pathogenicity: 4
  rcv: 6747
- chrom_no: 8
  start: 139570503
  stop: 139571539
  variation_type: 3
  pathogenicity: 4
  rcv: 7011
- chrom_no: 21
  start: 40742514
  stop: 40762546
  variation_type: 2
  pathogenicity: 2
  rcv: 536272
- chrom_no: 22
  start: 69080668
  stop: 69177002
  variation_type: 2
  pathogenicity: 4
  rcv: 539041
- chrom_no: 16
  start: 56769999
  stop: 56774226
  variation_type: 2
  pathogenicity: 2
  rcv: 554988
- chrom_no: 22
  start: 68890034
  stop: 69177002
  variation_type: 2
  pathogenicity: 4
  rcv: 633513
- chrom_no: 3
  start: 77116840
  stop: 77134716
  variation_type: 2
  pathogenicity: 2
  rcv: 638336
- chrom_no: 1
  start: 233194517
  stop: 233198703
  variation_type: 2
  pathogenicity: 2
  rcv: 638545
- chrom_no: 22
  start: 103031893
  stop: 103045546
  variation_type: 2
  pathogenicity: 4
  rcv: 640484
- chrom_no: 1
  start: 32358735
  stop: 32358736
  variation_type: 2
  pathogenicity: 4
  rcv: 6031
- chrom_no: 11
  start: 115120779
  stop: 115120779
  variation_type: 3
  pathogenicity: 4
  rcv: 8451
- chrom_no: 11
  start: 114836472
  stop: 114836472
  variation_type: 3
  pathogenicity: 4
  rcv: 8460
- chrom_no: 0
  start: 228345537
  stop: 228345537
  variation_type: 3
  pathogenicity: 3
  rcv: 256421
- chrom_no: 13
  start: 21775989
  stop: 21776009
  variation_type: 3
  pathogenicity: 3
  rcv: 256440
- chrom_no: 20
  start: 47545768
  stop: 47545769
  variation_type: 2
  pathogenicity: 3
  rcv: 256480
- chrom_no: 17
  start: 43479513
  stop: 43479513
  variation_type: 3
  pathogenicity: 3
  rcv: 416351
- chrom_no: 1
  start: 179454322
  stop: 179454323
  variation_type: 2
  pathogenicity: 2
  rcv: 256437
- chrom_no: 16
  start: 12899881
  stop: 12899882
  variation_type: 2
  pathogenicity: 4
  rcv: 5360
- chrom_no: 2
  start: 33100045
  stop: 33100046
  variation_type: 0
  pathogenicity: 3
  rcv: 1849900
- chrom_no: 13
  start: 99641905
  stop: 99641906
  variation_type: 2
  pathogenicity: 2
  rcv: 2265366
- chrom_no: 22
  start: 154156796
  stop: 154227925
  variation_type: 0
  pathogenicity: 3
  rcv: 852264
- chrom_no: 14
  start: 38632006
  stop: 38632007
  variation_type: 2
  pathogenicity: 4
  rcv: 622432
- chrom_no: 7
  start: 144900557
  stop: 144900557
  variation_type: 3
  pathogenicity: 4
  rcv: 622481
- chrom_no: 0
  start: 1718780
  stop: 1718782
  variation_type: 3
  pathogenicity: 4
  rcv: 623933

